agreement/player.go

```diff
@@ -266,6 +271,68 @@ func (p *player) handleCheckpointEvent(r routerHandle, e checkpointEvent) []acti
 		}}
 }
 
+// updateCredentialArrivalHistory is called at the end of a successful
+// uninterrupted round (just after ensureAction is generated) to collect
+// credential arrival times to dynamically set the filter timeout.
+func (p *player) updateCredentialArrivalHistory(r routerHandle, ver protocol.ConsensusVersion) {
+	// only append to lowestCredentialArrivals if this was a successful round completing in period 0.
+	if p.Period != 0 {
+		return
+	}
+	// look up the validatedAt time of the winning proposal-vote
+	re := readLowestEvent{T: readLowestVote, Round: p.Round, Period: p.Period}
+	re = r.dispatch(*p, re, proposalMachineRound, p.Round, p.Period, 0).(readLowestEvent)
+	if !re.Filled {
+		return
+	}
+
+	p.lowestCredentialArrivals = append(p.lowestCredentialArrivals, re.Vote.validatedAt)
+
+	// trim history to the last proto.DynamicFilterCredentialArrivalHistory samples.
+	proto := config.Consensus[ver]
+	if len(p.lowestCredentialArrivals) > proto.DynamicFilterCredentialArrivalHistory {
+		p.lowestCredentialArrivals = p.lowestCredentialArrivals[len(p.lowestCredentialArrivals)-proto.DynamicFilterCredentialArrivalHistory:]
+	}
+}
+
+// calculateFilterTimeout chooses the appropriate filter timeout.
+func (p *player) calculateFilterTimeout(ver protocol.ConsensusVersion, tracer *tracer) time.Duration {
+	proto := config.Consensus[ver]
+	if proto.DynamicFilterCredentialArrivalHistory <= 0 || p.Period != 0 {
+		// Either dynamic filter timeout is disabled, or we're not in period 0
+		// and therefore, can't use dynamic timeout
+		return FilterTimeout(p.Period, ver)
+	}
+	defaultTimeout := FilterTimeout(0, ver)
+	if proto.DynamicFilterCredentialArrivalHistory > len(p.lowestCredentialArrivals) {
+		// not enough samples, use the default
+		return defaultTimeout
+	}
+
+	sortedArrivals := make([]time.Duration, len(p.lowestCredentialArrivals))
+	copy(sortedArrivals[:], p.lowestCredentialArrivals[:])
+	sort.Slice(sortedArrivals, func(i, j int) bool { return sortedArrivals[i] < sortedArrivals[j] })
+	dynamicTimeout := sortedArrivals[proto.DynamicFilterTimeoutCredentialArrivalHistoryIdx] + proto.DynamicFilterTimeoutGraceInterval
+
+	// Make sure the dynamic filter timeout is not too small nor too large
+	if dynamicTimeout < proto.DynamicFilterTimeoutLowerBound {
+		dynamicTimeout = proto.DynamicFilterTimeoutLowerBound
+	}
+	if dynamicTimeout > defaultTimeout {
+		dynamicTimeout = defaultTimeout
+	}
+	tracer.log.Debugf("round %d, period %d: dynamicTimeout = %d, clamped timeout = %d", p.Round, p.Period, dynamicTimeout, timeout)
```

Comment:
I think this isn't logging what you want: it's getting the definition of `timeout` from `events.go` as an `eventType`.